include/tudocomp/coders/AdaptiveHuffmanCoder.hpp

```diff
@@ -0,0 +1,857 @@
+                }
+                return;
+            }
+
+            std::queue<bool> temp_left(do_code);
+            if ((pTree)->has_left()) {
+                temp_left.push(false);
+                gen_code(temp_left, code_write, symbol, (pTree)->left);
+            }
+
+            std::queue<bool> temp_right(do_code);
+            if ((pTree)->has_right()) {
+                temp_right.push(true);
+                gen_code(temp_right, code_write, symbol, (pTree)->right);
+            }
+
+            return;
+        }
+
+        /**
+         * \brief Creates the nyt-code for the current tree
+         * @param do_code Temporary queue
+         * @param write_code Result queue
+         * @param pTree Current node
+         */
+        inline void gen_nyt_code(std::queue<bool> &do_code, std::queue<bool> &write_code, node *pTree) {
+
+            if ((pTree)->weight == 0 && (pTree)->is_leaf()) {
+                while (!do_code.empty()) {
+                    bool next_bit = do_code.front();
+                    do_code.pop();
+                    write_code.push(next_bit);
+                }
+            }
+
+            std::queue<bool> temp_left(do_code);
+            if ((pTree)->has_left()) {
+                temp_left.push(false);
+                gen_nyt_code(temp_left, write_code, (pTree)->left);
+            }
+
+            std::queue<bool> temp_right(do_code);
+            if ((pTree)->has_right()) {
+                temp_right.push(true);
+                gen_nyt_code(temp_right, write_code, (pTree)->right);
+            }
+
+        }
+
+        /**
+         * Creates the standart encoding for a symbol
+         * @param code
+         * @param symbol
+         */
+        inline void gen_standard_code(std::queue<bool> &code, uliteral_t symbol) {
+            unsigned int symbol_size = sizeof(uliteral_t)*CHAR_BIT;
+            unsigned int bit_mask = pow(2, symbol_size - 1);
+
+            for (unsigned int i = 0; i < symbol_size; i++) {
+                if ((symbol & bit_mask) == bit_mask) {
+                    code.push(true);
+                } else {
+                    code.push(false);
+                }
+                symbol <<= 1;
+            }
+        }
+
+
+        inline void encode(codingtree *enc_codingtree, tdc::io::BitOStream &os, uliteral_t symbol) {
+            node *tree = enc_codingtree->root;
+            std::vector<bool> *dictionary = enc_codingtree->dictionary;
+
+            std::queue<bool> do_code, final_code;
+
+
+
+            // if node representing symbol exists
+            if ((*dictionary)[static_cast<unsigned int>(symbol)]) {
+                gen_code(do_code, final_code, symbol, tree);
+            } else {
+                // Node representing symbol does not exist
+                if (tree != nullptr) {
+                    gen_nyt_code(do_code, final_code, tree);
+                }
+                gen_standard_code(final_code, symbol);
+            }
+
+            bool sign;
```

Comment:
This is really insignificant, but this variable might as well just be declared inside the loop.

(I was about to write that it might be uninitialized if the loop never runs its body, but then noticed that it just gets read inside of it)